applications/CompressiblePotentialFlowApplication/custom_processes/define_3d_wake_process.cpp

```diff
@@ -0,0 +1,888 @@
+                                         const double SheddedWakeDistance,
+                                         const double SheddedWakeElementSize)
+    : Process(),
+      mrTrailingEdgeModelPart(rTrailingEdgeModelPart),
+      mrBodyModelPart(rBodyModelPart),
+      mrStlWakeModelPart(rStlWakeModelPart),
+      mTolerance(Tolerance),
+      mWakeNormal(rWakeNormal),
+      mWakeDirection(rWakeDirection),
+      mSwitchWakeDirection(SwitchWakeDirection),
+      mCountElementsNumber(CountElementsNumber),
+      mWriteElementsIdsToFile(WriteElementsIdsToFile),
+      mShedWakeFromTrailingEdge(ShedWakeFromTrailingEdge),
+      mSheddedWakeDistance(SheddedWakeDistance),
+      mSheddedWakeElementSize(SheddedWakeElementSize)
+{
+    KRATOS_ERROR_IF(mWakeNormal.size() != 3)
+        << "The mWakeNormal should be a vector with 3 components!"
+        << std::endl;
+}
+
+void Define3DWakeProcess::ExecuteInitialize()
+{
+    ModelPart& root_model_part = mrBodyModelPart.GetRootModelPart();
+    auto& r_nodes = root_model_part.Nodes();
+    VariableUtils().SetNonHistoricalVariable(UPPER_SURFACE, false, r_nodes);
+    VariableUtils().SetNonHistoricalVariable(LOWER_SURFACE, false, r_nodes);
+
+    InitializeTrailingEdgeSubModelpart();
+
+    InitializeWakeSubModelpart();
+
+    // Compute span direction as the cross product: mWakeNormal x mWakeDirection
+    MathUtils<double>::CrossProduct(mSpanDirection, mWakeNormal, mWakeDirection);
+
+    MarkTrailingEdgeNodesAndFindWingtipNodes();
+
+    ComputeWingLowerSurfaceNormals();
+
+    ComputeAndSaveLocalWakeNormal();
+
+    if(mShedWakeFromTrailingEdge){
+        KRATOS_INFO("Define3DWakeProcess") << " Shedding wake from the trailing edge" << std::endl;
+        ShedWakeSurfaceFromTheTrailingEdge();
+    }
+
+    MarkWakeElements();
+
+    RecomputeNodalDistancesToWakeOrWingLowerSurface();
+
+    MarkKuttaElements();
+
+    SaveLocalWakeNormalInElements();
+
+    AddWakeNodesToWakeModelPart();
+
+    if(mCountElementsNumber){
+        CountElementsNumber();
+    }
+
+    if(mWriteElementsIdsToFile){
+        WriteElementIdsToFile();
+    }
+}
+
+// This function initializes the variables and removes all of the elements and
+// nodes of the trailing edge submodelpart
+void Define3DWakeProcess::InitializeTrailingEdgeSubModelpart() const
+{
+    ModelPart& root_model_part = mrBodyModelPart.GetRootModelPart();
+    if(root_model_part.HasSubModelPart("trailing_edge_elements_model_part"))
+    {
+        // Clearing the variables and elements of the already existing
+        // trailing_edge_sub_model_part
+        ModelPart& trailing_edge_sub_model_part =
+            root_model_part.GetSubModelPart("trailing_edge_elements_model_part");
+
+        for (auto& r_element : trailing_edge_sub_model_part.Elements()){
+            r_element.SetValue(TRAILING_EDGE, false);
+            r_element.SetValue(KUTTA, false);
+            r_element.Reset(STRUCTURE);
+            r_element.Set(TO_ERASE, true);
+        }
```

Comment:
parallel utils

Reply:
The trailing_edge_sub_model_part typically has 100 to 1000 elements, when parallelizing this loop I noticed it runs slower...